python_package/madflow/lhe_writer.py

```diff
@@ -157,57 +200,61 @@ def lhe_parser(self, all_ps, res):
-            'scale': 0.0, # Q^2 scale for pdfs, get this from vegasflow?
-            'aqed': 0.0,  # alpha EW value    , get this from vegasflow?
-            'tag': '',
-            'comment': ''
-        } for wgt in res.numpy()]
-
-        index_to_pid = {
-            0: 2212, # p
-            1: 2212, # p
-            2: 6,    # t
-            3: -6    # t~
-        }
-
-        index_to_status = {
-            0: -1, # incoming particle
-            1: -1, # incoming particle
-            2:  1, # outgoing particle
-            3:  1, # outgoing particle
-        }
-
-        # we are missing the virtual particles
-        particles_info = [
-            [{
-            'pid': index_to_pid[i],
-            'status': index_to_status[i],
-            'mother1': 0,
-            'mother2': 0,
-            'color1': 0,
-            'color2': 0,
-            'E': ps[0],
-            'px': ps[1],
-            'py': ps[2],
-            'pz': ps[3],
-            'mass': np.sqrt(ps[0]**2 - ps[1]**2 - ps[2]**2 - ps[3]**2), # vectorize this?
-            'vtim': 0,
-            'helicity': 0,
-            } for i, ps in enumerate(ps_external)
-        ] for ps_external in all_ps.numpy()]
+        all_lumis = all_lumis / tf.reduce_sum(all_lumis, axis=1, keepdims=True)
+
+        events_info = []
+        particles_info = []
+        for wgt, ps_external, lumis, info in zip(
+            res.numpy(), all_ps.numpy(), all_lumis.numpy(), evt_info.numpy()
+        ):
+            pids, status_vec = self.lumis_to_pids(lumis)
+            nexternal = len(status_vec)
+            info = EventInfoTuple(*info)
+            events_info.append(
+                {
+                    "nexternal": nexternal,
+                    "ievent": 1,
+                    "wgt": wgt,
+                    "scale": info.scale,
+                    "aqed": 0.0,  # alpha EW value , get this from vegasflow?
+                    "aqcd": info.aqcd,
+                    "tag": "",
+                    "comment": "",
+                }
+            )
+
+            # we are missing the virtual particles
+            particles_info.append(
+                [
+                    {
+                        "pid": pid,
+                        "status": status,
+                        "mother1": 0,
+                        "mother2": 0,
+                        "color1": 0,
+                        "color2": 0,
+                        "E": ps[0],
+                        "px": ps[1],
+                        "py": ps[2],
+                        "pz": ps[3],
+                        "mass": np.sqrt(
+                            ps[0] ** 2 - ps[1] ** 2 - ps[2] ** 2 - ps[3] ** 2
+                        ),  # vectorize this?
+                        "vtim": 0,
+                        "helicity": 0,
+                    }
+                    for pid, status, ps in zip(pids, status_vec, ps_external)
+                ]
+            )
```

Comment:
If you move these loops to `self.dump_events` (which is call asynchronously) and then just pass the information you need, you get a huge speed up as the dumping is performed while the next iteration is running (so effectively only the last one will maybe have to wait for a little longer).